File: sdkwork-rtc-sdk-csharp/src/SDKWork.Rtc.Sdk/RtcProviderCatalog.cs
namespace Sdkwork.Rtc.Sdk;

using System.Linq;

public sealed record RtcProviderCatalogEntry(
    string providerKey,
    string pluginId,
    string driverId,
    bool defaultSelected
);

public static class RtcProviderCatalog
{
    public const string DEFAULT_RTC_PROVIDER_KEY = "volcengine";

    public static readonly IReadOnlyList<RtcProviderCatalogEntry> Entries =
    [
        new("volcengine", "rtc-volcengine", "sdkwork-rtc-driver-volcengine", true),
        new("aliyun", "rtc-aliyun", "sdkwork-rtc-driver-aliyun", false),
        new("tencent", "rtc-tencent", "sdkwork-rtc-driver-tencent", false),
        new("agora", "rtc-agora", "sdkwork-rtc-driver-agora", false),
        new("zego", "rtc-zego", "sdkwork-rtc-driver-zego", false),
        new("livekit", "rtc-livekit", "sdkwork-rtc-driver-livekit", false),
        new("twilio", "rtc-twilio", "sdkwork-rtc-driver-twilio", false),
        new("jitsi", "rtc-jitsi", "sdkwork-rtc-driver-jitsi", false),
        new("janus", "rtc-janus", "sdkwork-rtc-driver-janus", false),
        new("mediasoup", "rtc-mediasoup", "sdkwork-rtc-driver-mediasoup", false),
    ];

public static RtcProviderCatalogEntry? GetRtcProviderByProviderKey(string providerKey) =>
        Entries.FirstOrDefault(entry => entry.providerKey == providerKey);

}

File: sdkwork-rtc-sdk-flutter/lib/src/rtc_provider_catalog.dart
final class RtcProviderCatalogEntry {
  const RtcProviderCatalogEntry({
    required this.providerKey,
    required this.pluginId,
    required this.driverId,
    required this.defaultSelected,
  });

  final String providerKey;
  final String pluginId;
  final String driverId;
  final bool defaultSelected;
}

final class RtcProviderCatalog {
  static const String DEFAULT_RTC_PROVIDER_KEY = "volcengine";

  static const List<RtcProviderCatalogEntry> entries = <RtcProviderCatalogEntry>[
    RtcProviderCatalogEntry(
      providerKey: "volcengine",
      pluginId: "rtc-volcengine",
      driverId: "sdkwork-rtc-driver-volcengine",
      defaultSelected: true,
    ),
    RtcProviderCatalogEntry(
      providerKey: "aliyun",
      pluginId: "rtc-aliyun",
      driverId: "sdkwork-rtc-driver-aliyun",
      defaultSelected: false,
    ),
    RtcProviderCatalogEntry(
      providerKey: "tencent",
      pluginId: "rtc-tencent",
      driverId: "sdkwork-rtc-driver-tencent",
      defaultSelected: false,
    ),
    RtcProviderCatalogEntry(
      providerKey: "agora",
      pluginId: "rtc-agora",
      driverId: "sdkwork-rtc-driver-agora",
      defaultSelected: false,
    ),
    RtcProviderCatalogEntry(
      providerKey: "zego",
      pluginId: "rtc-zego",
      driverId: "sdkwork-rtc-driver-zego",
      defaultSelected: false,
    ),
    RtcProviderCatalogEntry(
      providerKey: "livekit",
      pluginId: "rtc-livekit",
      driverId: "sdkwork-rtc-driver-livekit",
      defaultSelected: false,
    ),
    RtcProviderCatalogEntry(
      providerKey: "twilio",
      pluginId: "rtc-twilio",
      driverId: "sdkwork-rtc-driver-twilio",
      defaultSelected: false,
    ),
    RtcProviderCatalogEntry(
      providerKey: "jitsi",
      pluginId: "rtc-jitsi",
      driverId: "sdkwork-rtc-driver-jitsi",
      defaultSelected: false,
    ),
    RtcProviderCatalogEntry(
      providerKey: "janus",
      pluginId: "rtc-janus",
      driverId: "sdkwork-rtc-driver-janus",
      defaultSelected: false,
    ),
    RtcProviderCatalogEntry(
      providerKey: "mediasoup",
      pluginId: "rtc-mediasoup",
      driverId: "sdkwork-rtc-driver-mediasoup",
      defaultSelected: false,
    )
  ];

  const RtcProviderCatalog._();
}

RtcProviderCatalogEntry? getRtcProviderByProviderKey(String providerKey) {
  for (final entry in RtcProviderCatalog.entries) {
    if (entry.providerKey == providerKey) {
      return entry;
    }
  }

  return null;
}
